tests/units/units.cpp

```diff
@@ -2799,3 +2831,593 @@ TEST(Units, equivalentUnitsMatchingBuiltinUnitsFromVariable)
+    ic->setImportSource(import);
+    ic->setImportReference("my_component");
+
+    importModel->component(0)->setMath(math);
+
+    libcellml::UnitsPtr u = libcellml::Units::create("fmol");
+    u->addUnit("farad");
+    u->addUnit("mole");
+
+    model->addComponent(ic);
+    model->addUnits(u);
+
+    auto flatModel = importer->flattenModel(model);
+
+    EXPECT_EQ(size_t(4), flatModel->unitsCount());
+    EXPECT_EQ("fmol", flatModel->units(0)->name());
+    EXPECT_EQ("fmol_1", flatModel->units(1)->name());
+    EXPECT_EQ("per_fmol", flatModel->units(2)->name());
+    EXPECT_EQ("fmol_1", flatModel->units(2)->unitAttributeReference(0));
+    EXPECT_EQ("per_fmol", flatModel->component(0)->variable(0)->units()->name());
+    EXPECT_TRUE(flatModel->component(0)->math().find("fmol_1") != std::string::npos);
+}
+
+void setupFlattenImportedUnits(libcellml::ImporterPtr &importer, libcellml::ModelPtr &model, const std::string &variation)
+{
+    libcellml::ModelPtr unitsDefinitionsModel;
+    libcellml::ImportSourcePtr import = createBaseModelImport(unitsDefinitionsModel, "importer/units/units_definitions.cellml");
+
+    libcellml::ModelPtr intermediaryImportModel1 = libcellml::Model::create("import_model1");
+    libcellml::ModelPtr intermediaryImportModel2 = libcellml::Model::create("import_model2");
+
+    importer = libcellml::Importer::create();
+
+    importer->addModel(unitsDefinitionsModel, "units_definitions");
+    importer->addModel(intermediaryImportModel1, "int1");
+    importer->addModel(intermediaryImportModel2, "int2");
+
+    model = libcellml::Model::create("main_model");
+
+    libcellml::UnitsPtr iu1 = libcellml::Units::create("per_fmol");
+    iu1->setImportSource(import);
+    iu1->setImportReference("per_fmol");
+
+    libcellml::UnitsPtr iu2 = libcellml::Units::create("fmol");
+    iu2->setImportSource(import);
+    iu2->setImportReference("fmol");
+
+    intermediaryImportModel1->addUnits(iu1);
+    intermediaryImportModel2->addUnits(iu2);
+
+    libcellml::ImportSourcePtr importIntermediary1 = libcellml::ImportSource::create();
+    importIntermediary1->setUrl("I_am_a_url");
+    importIntermediary1->setModel(intermediaryImportModel1);
+
+    libcellml::ImportSourcePtr importIntermediary2 = libcellml::ImportSource::create();
+    importIntermediary2->setUrl("I_am_a_url");
+    importIntermediary2->setModel(intermediaryImportModel2);
+
+    libcellml::UnitsPtr iu3 = libcellml::Units::create("per_fmol");
+    iu3->setImportSource(importIntermediary1);
+    iu3->setImportReference("per_fmol");
+
+    libcellml::UnitsPtr iu4 = libcellml::Units::create("fmol");
+    iu4->setImportSource(importIntermediary2);
+    iu4->setImportReference("fmol");
+
+    libcellml::UnitsPtr u = libcellml::Units::create("fmol");
+    u->addUnit("mole", "femto");
+    if (variation == "int1") {
+        intermediaryImportModel1->addUnits(u);
+    } else if (variation == "main") {
+        model->addUnits(u);
+    }
```

Comment:
Might want to have:
```c++
if (!variation.empty()) {
    ...
}
```

Reply:
Not sure what is wanted here, `setupFlattenImportedUnits` is called three times.  The variants are:
* "" (empty string)
* int1
* main

Reply:
I meant to have:
```c++
if (!variation.empty()) {
    libcellml::UnitsPtr u = libcellml::Units::create("fmol");
    u->addUnit("mole", "femto");
    if (variation == "int1") {
        intermediaryImportModel1->addUnits(u);
    } else if (variation == "main") {
        model->addUnits(u);
    }
}
```
i.e. only execute that code for the case where `variation` is not empty indeed.

Reply:
[877dc26](https://github.com/cellml/libcellml/pull/1167/commits/877dc26214a070ee094855f0a45c3c4122be0b15)